architecture/replication-layer.md

```diff
@@ -0,0 +1,77 @@
+
+#### Logs
+
+The log voted on by the Raft group contains an ordered set of commands that a quorum of replicas acknowledged and can be replayed to bring the node from a past state to its current state.
+
+This log also lets nodes that temporarily went offline to be "caught up" to the current state without needing to receive a copy of the existing data in the form of a snapshot.
+
+### Snapshots
+
+Each replica can be "snapshotted", which copies all of its data as of a specific MVCC timestamp. This snapshot can be sent to other nodes during a rebalance event to expedite replication.
+
+After receiving a snapshot, the node gets it up to date by playing all of the actions from the Raft group's committed log that have occurred since the snapshot was taken.
+
+### Leases
+
+Leases are held by a single member of the Raft group (usually the same node as the Raft group leader), and the leaseholder is responsible for receiving `BatchRequests` for all reads and writes related to keys in the range. This has the benefit of improving the speed of reads drastically (which don't need consensus, since a quorum of nodes already agreed with the leaseholder as to the status of the node after the write ).
+
+If there is no Lease holder, any range receiving a request will attempt to become to Lease holder.
+
+Since reads bypass Raft, a new lease holder will, among other things, ascertain that its timestamp cache does not report timestamps smaller than the previous lease holder's (so that it's compatible with reads which may have occurred on the former lease holder). This is accomplished by letting leases enter a stasis period (which is just the expiration minus the maximum clock offset) before the actual expiration of the lease, so that all the next lease holder has to do is set the low water mark of the timestamp cache to its new lease's start time.
+
+As a lease enters its stasis period, no more reads or writes are served, which is undesirable. However, this would only happen in practice if a node became unavailable. In almost all practical situations, no unavailability results since leases are usually long-lived (and/or eagerly extended, which can avoid the stasis period) or proactively transferred away from the lease holder, which can also avoid the stasis period by promising not to serve any further reads until the next lease goes into effect.
+
+#### Co-location with Raft leadership
+
+The range lease is completely separate from Raft leadership, and so without further efforts, Raft leadership and the Range lease might not be held by the same Replica. Since it's expensive to not have these two roles colocated (the lease holder has to forward each proposal to the leader, adding costly RPC round-trips), each lease renewal or transfer also attempts to colocate them. In practice, that means that the mismatch is rare and self-corrects quickly.
+
+### Membership Changes: Rebalance/Repair
```

Comment:
Also, for range rebalancing:
- This [blog post](https://docs.google.com/document/d/1PsBLlKi-_Ycl56pB5DSZ4tkEAzAbsthenD2sb19pTTc/edit) @BramGruneir and I are working on might have some good details to reuse in terms of the heuristics around range rebalancing. 
- In 1.1, range size is now a primary consideration in range rebalancing: https://github.com/cockroachdb/docs/issues/1560. Would be cool if we could integrate that (which might mean having a 1.0 version and a 1.1 version of these pages, like for our usage docs).